internal/venice-client-common/src/main/java/com/linkedin/venice/compression/GzipCompressor.java

```diff
@@ -12,17 +12,33 @@
 
 
 public class GzipCompressor extends VeniceCompressor {
+  GzipPool gzipPool;
+
   public GzipCompressor() {
     super(CompressionStrategy.GZIP);
+    this.gzipPool = new GzipPool();
   }
 
   @Override
   public byte[] compress(byte[] data) throws IOException {
     ByteArrayOutputStream bos = new ByteArrayOutputStream();
```

Comment:
I wonder if we should make this `BAOS` also thread-local, and reset it for reuse. The BAOS can grow several times during deflation, which would result in a copy each time, so it seems potentially valuable to hang on to it. I think in order to make this work correctly we should move the return statement to the end of the try block, and then add `bos.reset()` to the finally block (i.e. the reset should happen after the `bos.toByteArray()`, IIUC).

Reply:
So what I'm saying in that last idea is that this function could end up looking like this:

```
  public byte[] compress(byte[] data) throws IOException {
    ReusableGzipOutputStream out = gzipPool.forStream(bos);
     try {
       out.writeHeader();
       out.write(data);
       out.finish();
       return out.toByteArray();
     } finally {
       out.reset();
     }
  }
```

Reply:
Honestly did not fully understand the free-writing :)    I updated the PR with reusing a BOS. Not sure if you meant something along that line.

Reply:
Thanks, looks great! Let's do one more CI run just to make sure and I think we should merge this in the current state if CI looks good.